From: Shiji Yang <yangshiji66@outlook.com>
Date: Thu, 18 Jan 2024 10:10:39 +0000
Subject: [PATCH] mtd: spi-nor: introduce segmented read hack

The ath79 SoC GPIO voltage is 2.5V. In order to be compatible
with 3.3V NOR Flash chips, most devices use resistors for voltage
matching. However, some devices (such as TP-Link WR2543N) use a
coupling capacitor to connect Flash and GPIO. This has led to a
strange behavior where the SPI bus is unable to continuously read
correct data from Flash chips. This patch workaround this issue by
limiting the data size of each SPI transfer cycle and inserting
some delay time between each read. User can add the newly designed
property 'read-segment-size' to the dts flash node to enable this
feature.

Signed-off-by: Shiji Yang <yangshiji66@outlook.com>
---

--- a/drivers/mtd/spi-nor/core.c
+++ b/drivers/mtd/spi-nor/core.c
@@ -2089,6 +2089,7 @@ static int spi_nor_read(struct mtd_info
 			size_t *retlen, u_char *buf)
 {
 	struct spi_nor *nor = mtd_to_spi_nor(mtd);
+	size_t read_size = nor->segment_size;
 	loff_t from_lock = from;
 	size_t len_lock = len;
 	ssize_t ret;
@@ -2102,9 +2103,17 @@ static int spi_nor_read(struct mtd_info
 	while (len) {
 		loff_t addr = from;
 
+		if (unlikely(nor->flags & SNOR_F_SEGMENTED_READ)) {
+			if (read_size > len)
+				read_size = len;
+			usleep_range(10, 20);
+		} else {
+			read_size = len;
+		}
+
 		addr = spi_nor_convert_addr(nor, addr);
 
-		ret = spi_nor_read_data(nor, addr, len, buf);
+		ret = spi_nor_read_data(nor, addr, read_size, buf);
 		if (ret == 0) {
 			/* We shouldn't see 0-length reads */
 			ret = -EIO;
@@ -2856,6 +2865,9 @@ static void spi_nor_init_flags(struct sp
 	if (of_property_read_bool(np, "no-wp"))
 		nor->flags |= SNOR_F_NO_WP;
 
+	if (!of_property_read_u32(np, "read-segment-size", &nor->segment_size))
+		nor->flags |= SNOR_F_SEGMENTED_READ;
+
 	if (flags & SPI_NOR_SWP_IS_VOLATILE)
 		nor->flags |= SNOR_F_SWP_IS_VOLATILE;
 
--- a/drivers/mtd/spi-nor/core.h
+++ b/drivers/mtd/spi-nor/core.h
@@ -133,6 +133,7 @@ enum spi_nor_option_flags {
 	SNOR_F_RWW		= BIT(14),
 	SNOR_F_ECC		= BIT(15),
 	SNOR_F_NO_WP		= BIT(16),
+	SNOR_F_SEGMENTED_READ	= BIT(17),
 };
 
 struct spi_nor_read_command {
--- a/include/linux/mtd/spi-nor.h
+++ b/include/linux/mtd/spi-nor.h
@@ -406,6 +406,7 @@ struct spi_nor {
 	enum spi_nor_protocol	reg_proto;
 	bool			sst_write_second;
 	u32			flags;
+	u32			segment_size;
 	enum spi_nor_cmd_ext	cmd_ext_type;
 	struct sfdp		*sfdp;
 	struct dentry		*debugfs_root;
